launcher/internal/gpu/driverinstaller.go

```diff
@@ -0,0 +1,162 @@
+// https://pkg.go.dev/cos.googlesource.com/cos/tools.git@v0.0.0-20241008015903-8431fe581b1f/src/cmd/cos_gpu_installer#section-readme
+// README specifies docker command where this function uses containerd for launching and managing the gpu driver installer container.
+func (di *DriverInstaller) InstallGPUDrivers(ctx context.Context) error {
+	err := remountAsExecutable(InstallationHostDir)
+	if err != nil {
+		return fmt.Errorf("failed to remount the installation directory: %v", err)
+	}
+
+	gpuType, err := deviceinfo.GetGPUTypeInfo()
+	if err != nil {
+		return fmt.Errorf("failed to get the gpu type info: %v", err)
+	}
+
+	if !gpuType.OpenSupported() {
+		return fmt.Errorf("unsupported gpu type %s, please retry with one of the supported gpu types: %v", gpuType.String(), supportedGpuTypes)
+	}
+
+	ctx = namespaces.WithNamespace(ctx, namespaces.Default)
+	installerImageRef, err := getInstallerImageReference()
+	if err != nil {
+		di.logger.Printf("failed to get the installer container image reference: %v", err)
+		return err
+	}
+
+	di.logger.Printf("cos gpu installer version : %s", installerImageRef)
+	image, err := di.cdClient.Pull(ctx, installerImageRef, containerd.WithPullUnpack)
+	if err != nil {
+		return fmt.Errorf("failed to pull installer image: %v", err)
+	}
+
+	mounts := []specs.Mount{
+		{
+			Type:        "volume",
+			Source:      "/dev",
+			Destination: "/dev",
+			Options:     []string{"rbind", "rw"},
+		}, {
+			Type:        "volume",
+			Source:      "/",
+			Destination: "/root",
+			Options:     []string{"rbind", "rw"},
+		},
+	}
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		di.logger.Printf("cannot get hostname: %v", err)
+	}
+
+	container, err := di.cdClient.NewContainer(
+		ctx,
+		installerContainerID,
+		containerd.WithImage(image),
+		containerd.WithNewSnapshot(installerSnapshotID, image),
+		containerd.WithNewSpec(oci.WithImageConfig(image),
+			oci.WithPrivileged,
+			oci.WithProcessArgs("/cos-gpu-installer", "install", "-version=default", fmt.Sprintf("-host-dir=%s", InstallationHostDir)),
+			oci.WithAllDevicesAllowed,
+			oci.WithHostDevices,
+			oci.WithMounts(mounts),
+			oci.WithHostNamespace(specs.NetworkNamespace),
+			oci.WithHostNamespace(specs.PIDNamespace),
+			oci.WithHostHostsFile,
+			oci.WithHostResolvconf,
+			oci.WithEnv([]string{fmt.Sprintf("HOSTNAME=%s", hostname)})))
+	if err != nil {
+		return fmt.Errorf("failed to create gpu driver installer container: %v", err)
+	}
+
+	task, err := container.NewTask(ctx, cio.NewCreator(cio.WithStdio))
+	if err != nil {
+		return fmt.Errorf("failed to create gpu driver installation task: %v", err)
+	}
+
+	statusC, err := task.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to wait for gpu driver installation task: %v", err)
+	}
+
+	if err := task.Start(ctx); err != nil {
+		return fmt.Errorf("failed to start gpu driver installation task: %v", err)
+	}
+
+	status := <-statusC
+	code, _, _ := status.Result()
```

Comment:
what if the returned code is not 0, do we just log the code and return nil? We should apply retry to non-zero code and return an error instead of nil.

Reply:
Make sense to log and return an error. Returned error would not be retryable as we do not want to retry on the gpu driver installation failure.